### DISCUSSION TITLE
md5 optimization at  IA

### DISCUSSION THREAD

**Discussion by moting9** [minio/minio#16231] [Show and tell]
Hello, Harshhavardhana, Klauspost
Recently, We made 2 optimizations at MD5.
The 1st MD5 patch is based on AVX512 ,  https://go-review.googlesource.com/c/go/+/455235
This is the perf improvement at 2nd generation Intel Xeon scalable server
![image](https://user-images.githubusercontent.com/24264611/207250825-c1011771-ba4d-4d4d-8b33-a72499f3ce97.png)

This is the perf improvement at 3rd generation Intel Xeon scalable server 
![image](https://user-images.githubusercontent.com/24264611/207251264-f6dec2c3-ab3b-4ad8-9543-08bb37893e8a.png)

The 2nd MD5 patch adjust asm https://go-review.googlesource.com/c/go/+/433475 and is simliar as  [CL 283538](https://golang.org/cl/283538)
This is the perf improvement at 2nd generation Intel Xeon scalable server
![image](https://user-images.githubusercontent.com/24264611/207254304-293e516f-200f-4ba0-ad64-00e331699e34.png)

This is the perf improvement at 3rd generation Intel Xeon scalable server 
![image](https://user-images.githubusercontent.com/24264611/207254400-c6d02e28-baff-434f-bed6-a25dde7ea9d3.png)

Based on comments of the 1st MD5 patch - https://go-review.googlesource.com/c/go/+/455235
the golang community seems opposed to AVX512 as well as in general in improvements to MD5. They consider it obsolete and want people to use as little as possible.  But MD5 is widely used in object storage and occupies high CPU usage when PUT. Is it possible that community to help state in the PR - https://go-review.googlesource.com/c/go/+/455235 they value a contribution like this?  It may be helpful because it tells the engineers at Google that the performance improvement is relevant and helpful to someone else other than the author, and may inspire them to accept the change.  Thanks!

The sha-ni patch https://go-review.googlesource.com/c/go/+/408795  has some community input recently, and the code review becomes active. 

Thanks.

**Comment by klauspost** [comment]
I am the author of [CL 283538](https://golang.org/cl/283538), so I don't have many hopes for your code to make it through, unless other people at Google suddenly needs it.

We have implemented the scalar improvements in our [md5-simd](https://github.com/minio/md5-simd), which also has multi-lane avx512 - but no single-lane avx512,

So rather than wait (probably forever)  for this to be accepted, we are using our own md5 software where applicable.
